api/resources_portal/views/attachment.py

```diff
@@ -0,0 +1,78 @@
+from rest_framework import serializers, viewsets
+from rest_framework.permissions import BasePermission, IsAdminUser, IsAuthenticated
+from rest_framework.response import Response
+
+from guardian.shortcuts import assign_perm
+
+from resources_portal.models import Attachment, MaterialRequest
+from resources_portal.views.relation_serializers import MaterialRelationSerializer
+
+
+class AttachmentSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Attachment
+        fields = (
+            "id",
+            "filename",
+            "description",
+            "s3_bucket",
+            "s3_key",
+            "deleted",
+            "created_at",
+            "updated_at",
+            "sequence_map_for",
+        )
+        read_only_fields = ("id", "created_at", "updated_at")
+
+
+class AttachmentDetailSerializer(AttachmentSerializer):
+    sequence_map_for = MaterialRelationSerializer()
+
+
+class HasViewAttachmentOrIsAdminUser(BasePermission):
+    def has_object_permission(self, request, view, obj):
+        return request.user.has_perm("view_attachment", obj) or request.user.is_staff
+
+
+class HasModifyAttachmentOrIsAdminUser(BasePermission):
+    def has_object_permission(self, request, view, obj):
+        return request.user.has_perm("modify_attachment", obj) or request.user.is_staff
+
+
+class AttachmentViewSet(viewsets.ModelViewSet):
+    queryset = Attachment.objects.all()
+
+    def get_serializer_class(self):
+        if self.action == "retrieve":
+            return AttachmentDetailSerializer
+
+        return AttachmentSerializer
+
+    def get_permissions(self):
+        if self.action == "list":
+            permission_classes = [IsAuthenticated, IsAdminUser]
+        elif self.action == "retrieve":
+            permission_classes = [IsAuthenticated, HasViewAttachmentOrIsAdminUser]
+        elif self.action == "create":
+            permission_classes = [IsAuthenticated]
+        else:
+            permission_classes = [IsAuthenticated, HasModifyAttachmentOrIsAdminUser]
+
+        return [permission() for permission in permission_classes]
+
+    def create(self, request, *args, **kwargs):
+        if not (
+            MaterialRequest.objects.filter(requester=request.user, is_active=True).exists()
+            or MaterialRequest.objects.filter(assigned_to=request.user, is_active=True).exists()
```

Comment:
This is quite a difficult thing to do because attachments aren't associated with an organization initially. The permissions to link an attachment to material request are already as you describe: anyone with the ability to approve requests can currently link attachments to the material request. The only permission I could maybe put here would be that the user has to be in any organization with an active material request, but that's pretty permissive as it is. Is that worth doing, or should we just let anyone upload an attachment? Or should we link attachments to an organization initially?

Reply:
> The only permission I could maybe put here would be that the user has to be in any organization with an active material request, but that's pretty permissive as it is.

Yeah it's pretty permissive, but I don't think that's an issue. I think it's better than forcing the assignee to be the one to upload things.

Reply:
Ok, I added this in. My code block for this is pretty ugly- if there's a way to simplify it, lmk. If not, there it is.